### ISSUE TITLE
Increase job timeout resolution / reject requests below resolution threshold

### ISSUE THREAD

**Issue by jwulf** [camunda/zeebe#5073]
At the moment, the broker job timeout resolution is 30 seconds. In the client library I can set the `timeout` for an ActivateJobs request that is less than 30 seconds.

However, when I run my program, I see retries happening every 30 seconds. Without understanding the internal mechanics of the broker, or reading the documentation to discover how it works, the behaviour is unexpected and inexplicable.

**Describe the solution you'd like**
Three potential approaches: 

1. Increase the broker timeout resolution to 1s. This may have a performance impact on the operation of the broker. This requires no change to the client libraries, and more closely aligns with an intuitive, unconscious expectation of users.
2. If not this, then code the client libraries so that the timeout has to be specified using a custom class that explicitly takes intervals of 30s - eg: `Job.Timeout30SecondBlock(1)`. This makes the client a little more complex, but the behaviour is semantically explicit. It makes getting started a little more complicated, but pushes the validation to the IDE and explicit and discoverable.
3. Throw an error when an ActivateJobs request is received with a timeout that is either less than 30 seconds, or not a multiple of 30 seconds. The error message would say: `Job timeout must be a multiple of 30 seconds`. This pushes the validation to run-time, but is explicit and discoverable. This behaviour means less impact on client libraries, and is easier to change in the future without requiring changes to the client libraries.

**Comment by sergeylebed** [comment]
I vote for solution #1

**Comment by jwulf** [comment]
Solution 1 would be best, but may be too far in the future. I would be happy with 3 now, and 1 later.

**Comment by npepinpe** [comment]
Let's go with the same way we implemented it for our timer catch events.

**Comment by Zelldon** [comment]
I just wanted to work on this, because it is on the ready queue. @saig0 mentioned that it might not be ideal since we currently rewrite the engine and he sees no need for it in this quarter. Any thoughts @npepinpe ?

**Comment by npepinpe** [comment]
I was under the assumption this wouldn't be a lot of work, but if you believe it is then we should postpone it until after the refactoring - it's not currently urgent.

**Comment by Zelldon** [comment]
I don't think that it is lot of work ^^

**Comment by npepinpe** [comment]
Would it add more complexity/work to the engine refactoring? @saig0

**Comment by npepinpe** [comment]
I thought about it some more, and it's probably a waste of time to start it now if we don't know how it'll impact the upcoming refactoring. Let's revisit it after.

**Comment by saig0** [comment]
I guess that it has not much impact on the refactoring. If Chris starts now then he is probably done before we start :smile:

**Comment by sstro** [comment]
Any chance this feature gets implemented sometime soon ? We have long-running jobs, and usually cannot predict for how long a job will run, so setting the timeout at the start is not an option. Thanks.

**Comment by saig0** [comment]
Currently, it is not planned for the near future. We're focusing on topics with a higher priority/impact. 

But this might be a good candidate for community contributions. :rocket:

**Comment by nhomble** [comment]
In use cases where teams want process instances to finish in less than 30 seconds, what options do we have? 

In a rolling deployment strategy, it's going to be common that the old instances will lock jobs that'll only be redistributed through this timeout.

Maybe I am missing something since this ticket's priority kept getting push, but this seems like a critical blocker for use cases where process instances need to complete in the 10s of seconds.

**Comment by korthout** [comment]
@nhomble I agree. Option 1 is the best solution for users as it allows for higher resolution of the job timeout checker. Performance degradation must be avoided on such a fundamental part of the engine.

How could we avoid degrading performance in the `JobTimeoutTrigger` (i.e. the checker)?
- [x] cache Job Timeout commands it appends to the log to avoid overloading the logstream
  - this is already done using caching in the Task Result Builder
- [ ] yield the actor to avoid blocking other checkers
  - this requires: stopping the iteration ([example](https://github.com/camunda/zeebe/blob/b1cbfdb5c4c677caf4c935045225d3ef8d12dce5/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageTimeToLiveChecker.java#L91)) once a particular number of items has been seen (this value should be configurable)
  - and requires: continuing the iteration from a specific point by keeping an index ([example](https://github.com/camunda/zeebe/blob/b1cbfdb5c4c677caf4c935045225d3ef8d12dce5/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageTimeToLiveChecker.java#L79-L83)) 

What else would we need to do?
- [ ] the checker interval must be configurable (this allows us to tweek things on saas or to reset back to 30s if it causes issues)
- [ ] the default checker interval can be changed to 100ms 

As the change itself is small, but some effort is required to prevent the checker from blocking other checkers, I'm sizing this as `medium`.

---

There are multiple reasons to resolve this issue:
- the hidden resolution is confusing to users
- low latency use cases have no workaround

Additionally, a customer has highlighted that there is no workaround in a support case.

I'm marking this as `priority:upcoming`. I'll raise it in the upcoming planning session.

**Comment by korthout** [comment]
As this blocks low-latency cases, I'm starting to feel that this should be considered a bug rather than a feature. The feature is that jobs can timeout. It's a bug that sub-30 second timeouts are not supported.

Curious what others think. When this would be changed, should it be backported to older versions?

**Comment by oleschoenburg** [comment]
I agree that this is something that we should change. Using 1 second as granularity feels right to me. I also think that we should backport it, at least for versions that also have command caching.

**Comment by aleksander-dytko** [comment]
ZPA planning: We'll prioritize this issue for the next iteration. 
Reasoning:
* We need to focus on the feature-related work 
* This issue is not urgent
* We won't be able to complete it for the upcoming patch release

**Comment by korthout** [comment]
The team has too many other topics that are considered even higher priority. We postpone this to another iteration, but we do want to work on this soon.

**Comment by korthout** [comment]
This could also make a good onboarding issue, as the problem and [possible solution](https://github.com/camunda/zeebe/issues/5073#issuecomment-1916609764) should be clear now.
